### PR TITLE
Stm32 pwm support

### PR DESCRIPTION
**Pull request by mphalke**
1. Added the STM32 PWM module initial implantation. The changes are tested using STM32F769I-DISCO board
2. Updated gpio module to modify the alternate gpio configuration

### CODE REVIEW

File: drivers/platform/stm32/stm32_pwm.c [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+	extra = (struct stm32_pwm_desc *)calloc(1, sizeof(*extra));
+	if (!extra) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	/* Initialize the PWM timer */
+	ret = _init_pwm_timer(extra, param);
+	if (ret)
```
Comment (amiclaus) [inline drivers/platform/stm32/stm32_pwm.c:193]: I think `free(extra)` should also be performed in the below error checks via a conditional jump.

File: drivers/platform/stm32/stm32_gpio.c [hunk 2]
```diff
@@ -107,6 +107,10 @@ static int32_t _gpio_init(struct gpio_desc *desc,
 #ifdef GPIOH
 	else if (pextra->port == GPIOH)
 		__HAL_RCC_GPIOH_CLK_ENABLE();
+#endif
+#ifdef GPIOJ
+	else if (pextra->port == GPIOJ)
+		__HAL_RCC_GPIOJ_CLK_ENABLE();
```
Comment (buha) [inline drivers/platform/stm32/stm32_gpio.c:113]: this commit has two independent changes, let's split it into two for better history tracking

Comment (mphalke) [reply drivers/platform/stm32/stm32_gpio.c:113]: sorry, can you help me with this. Looks like I have merged the master branch into this branch to even the remote and upstream branches and pushed a new commit related to pwm error handling, so the initial commits about gpio changes has moved down below merge changes from master (logs added below). I am not sure how can I remove the older commit Or do I need to remove gpio changes and push them again as two separate commits?

commit 97d1c372f100b71dd693bcf5b31a37b5fb1f54a7 (HEAD -> stm32-pwm-support, origin/stm32-pwm-support)
Author: mahphalke <Mahesh.Phalke@analog.com>
Date:   Thu Sep 16 13:27:32 2021 +0530

    Fixed the review comments related to pwm error handlings

    Signed-off-by: mahphalke <Mahesh.Phalke@analog.com>

commit afbdb3966090a3b60df2674d8f284fa23544b4b1
Merge: 59ac048a fa40d58a
Author: Mahesh Phalke <62383520+mphalke@users.noreply.github.com>
Date:   Thu Sep 16 13:17:41 2021 +0530

    Merge branch 'analogdevicesinc:master' into stm32-pwm-support

commit fa40d58aa19a2116731c4b568bf47b27f30daca7 (upstream/master, origin/master, origin/HEAD, master)
Author: Janani Sunil <janani.sunil@analog.com>
Date:   Wed Aug 25 14:27:25 2021 +0530

    drivers:temperature:adt7420:Add new supported devices of adt7xx family

    Signed-off-by: Janani Sunil <janani.sunil@analog.com>

commit f392b59a4756b8f4e50b54ae846f20848ed0eec1 (upstream/testbranch)
Author: APorumb <andrei.porumb@analog.com>
Date:   Fri Sep 10 10:23:19 2021 +0300

    iio_adpd1080 build fix using new irq_platform_ops

**Comment by mphalke** [comment]
Removed the out of scope commit (merge from master). Please review the changes

**Review by amiclaus — APPROVED** [review]

**Review by amiclaus — CHANGES_REQUESTED** [review]
Actually: https://github.com/analogdevicesinc/no-OS/pull/1094/commits/886ea7b292d02c137fefc1628580c7909a3fa9dc

You shouldn't do a separate commit with the review fixes. So please squash it with: https://github.com/analogdevicesinc/no-OS/pull/1094/commits/3ce992aab44ca9d5b8156f5132bbfe661f43f761

**Comment by mphalke** [comment]
Squashed the 1st and 2nd commits

File: drivers/platform/stm32/stm32_pwm.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/**
+ * @struct stm32_pwm_init_param
+ * @brief Structure holding the STM32 PWM parameters.
+ */
+typedef struct stm32_pwm_init_param {
```
Comment (cristiantpop) [inline drivers/platform/stm32/stm32_pwm.h:65]: Please don't use typedef

File: drivers/platform/stm32/stm32_pwm.h [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	gpio_init_param *gpio;
+	/** Get timer clock frequency function */
+	uint32_t (*get_timer_clock)(void);
+} stm32_pwm_init_param;
+
+/**
+ * @struct stm32_pwm_desc
+ * @brief Structure holding the STM32 PWM descriptor.
+ */
+typedef struct stm32_pwm_desc {
```
Comment (cristiantpop) [inline drivers/platform/stm32/stm32_pwm.h:81]: Please don't use typedef

File: drivers/platform/stm32/stm32_pwm.h [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	/** PWM Timer Channel */
+	uint32_t timer_chn;
+} stm32_pwm_desc;
+
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+
+/* Initialize the PWM generator device */
+int32_t stm32_pwm_init(struct pwm_desc **desc,
```
Comment (cristiantpop) [inline drivers/platform/stm32/stm32_pwm.h:97]: I think that you already have declarations of this functions in
https://github.com/analogdevicesinc/no-OS/blob/master/include/pwm.h
please implement those.

**Review by cristiantpop — CHANGES_REQUESTED** [review]

File: drivers/platform/stm32/stm32_pwm.h (hunk 3 above)
Comment (buha) [reply drivers/platform/stm32/stm32_pwm.h:65]: yeah, so an explanation for Mahesh here, we kind of follow the linux coding guidelines (available online) and they have a point there suggesting typedef can be confusing, hence we don't use it either for new code

Comment (mphalke) [reply drivers/platform/stm32/stm32_pwm.h:65]: Thanks for the explanation. I shall remove the typedef from structures.

File: drivers/platform/stm32/stm32_pwm.h (hunk 5 above)
Comment (mphalke) [reply drivers/platform/stm32/stm32_pwm.h:97]: I have only implemented the functions which are necessary/required for STM32 PWM init. Do you expect to implement all those declared in pwm.h?

Comment (cristiantpop) [reply drivers/platform/stm32/stm32_pwm.h:97]: Implementing only the functions that you need is OK, the driver can be extended if it is necessary.

**Comment by mphalke** [comment]
Hi Reviewers,
Is this PR good to merge?

**Comment by buha** [comment]
We now have https://github.com/analogdevicesinc/no-OS/blob/master/include/no_os_pwm.h pwm interface so the correct aproach would be to implement something that follows along that API.

**Pull request closed without merging**